### PR TITLE
I/6502: Refactor TableUtils.removeRows() logic.

### PR DESCRIPTION
**Pull request by jodator**
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Fix: Remove row in complex tables now properly move cells from removed rows. Closes ckeditor/ckeditor5#6502.

---

### Additional information

This PR also includes fixes for some issues when not operating on `TableEditing` model (no post-fixer). https://github.com/ckeditor/ckeditor5/issues/6574.

The actual fix is commented below but to fix that I had to understand the logic once again. Thus refactoring to removing all rows at once (this bring also a benefit of fewer operations when removing more than one row.

As a side effect, I've fixed some of the `removeRows()` test as it operated on wrongly created table. See: https://github.com/ckeditor/ckeditor5/issues/6574.

### CODE REVIEW

File: src/tableutils.js [hunk 1]
```diff
@@ -730,60 +762,60 @@ function breakSpanEvenly( span, numberOfCells ) {
...
-				writer.createPositionAt( table.getChild( row ), 0 );
+				writer.createPositionAt( row, 0 );
+
 			writer.move( writer.createRangeOn( cellToMove ), targetPosition );
-			updateNumericAttribute( 'rowspan', rowspanToSet, cellToMove, writer );
+			updateNumericAttribute( 'rowspan', rowspan, cellToMove, writer );
+
 			previousCell = cellToMove;
-		} else {
+		} else if ( !isSpanned ) {
```
Comment (jodator) [inline src/tableutils.js:816]: This is the line that caused a problem in previous implementation.

File: tests/tableutils.js [hunk 2]
```diff
@@ -826,6 +937,21 @@ describe( 'TableUtils', () => {
...
+					[ '20', '23', '24' ]
+				] ) );
+
+				tableUtils.removeRows( root.getChild( 0 ), { at: 1 } );
+
+				assertEqualMarkup( getData( model, { withoutSelection: true } ), modelTable( [
+					[ '00', { rowspan: 2, contents: '01' }, '02', '03', '04' ],
+					[ '20', '12', '23', '24' ]
+				] ) );
+			} );
```
Comment (niegowski) [inline tests/tableutils.js:954]: This test gives strange results in manual test:
Zrzut ekranu 2020-04-15 o 13 25 16

File: tests/tableutils.js [hunk 3]
```diff
@@ -924,17 +1050,37 @@ describe( 'TableUtils', () => {
...
 				setData( model, modelTable( [
-					[ '00', { contents: '01', rowspan: 3 } ],
-					[ '10' ],
-					[ '20' ]
+					[ '00', '01', '02', '03' ],
+					[ { rowspan: 4, contents: '10' }, { rowspan: 3, contents: '11' }, { rowspan: 2, contents: '12' }, '13' ],
+					[ { rowspan: 3, contents: '23' } ],
+					[ '32' ],
+					[ '41', '42' ]
 				] ) );
```
Comment (niegowski) [inline tests/tableutils.js:1060]: This gives strange results in manual test: 
Zrzut ekranu 2020-04-15 o 13 31 23
Should be looking like this:
```
+----+----+----+----+
| 00 | 01 | 02 | 03 |
+----+----+----+----+
| 10 | 11 | 12 | 13 |
+    +    +    +----+
|    |    |    | 23 |
+    +    +----+    +
|    |    | 32 |    |
+    +----+----+    +
|    | 41 | 42 |    |
+----+----+----+----+
```
 
Apr-15-2020 13-35-15
This seems to be error in some other place (maybe post-fixers?).

Comment (jodator) [reply tests/tableutils.js:1060]: Welcome to the table world. This is how browsers renders tables :shrug:. I didn't find the talk with some external dev on this but AFAIR we cannot do anything here.

The problem is with "collapsing rows" - just look at the HTML markup - it is OK.

File: tests/tableutils.js (hunk 2 above)
Comment (jodator) [reply tests/tableutils.js:954]: See comment below - it is how browsers renders table :man\_shrugging:

**Comment by jodator** [comment]
> Comments below, also some missing scenario:

New/old bug - unrelated to the change. Also a corner case for me and requires much more tests. I'll create a follow up for this issue.

File: tests/tableutils.js (hunk 2 above)
Comment (niegowski) [reply tests/tableutils.js:954]: This is different case (it changes order of cells `12` and `23`).

Comment (jodator) [reply tests/tableutils.js:954]: To double check:



WFM but after changes from your other comment.

File: src/tableutils.js [hunk 4]
```diff
@@ -267,13 +267,13 @@ export default class TableUtils extends Plugin {
...
-	 *		  3 │   │ h │ i │
+	 *		  1 │   │ d │ e │  <-- remove from here  │   │ d │ i │ 1
+	 *		    │   │   ├───┤        will give:      ├───┼───┼───┤
+	 *		  2 │   │   │ f │                        │ j │ k │ l │ 2
+	 *		    │   │   ├───┤                        └───┴───┴───┘
+	 *		  3 │   │   │ g │
 	 *		    ├───┼───┼───┤
-	 *		  4 │ j │ k │ l │
+	 *		  4 │ h │ i │ j │
 	 *		    └───┴───┴───┘
```
[the remark this answers is not in the document]
Comment (jodator) [reply src/tableutils.js]: Fixed.

File: src/tableutils.js [hunk 5]
```diff
@@ -283,27 +283,59 @@ export default class TableUtils extends Plugin {
...
+				cellsToMove.set( column, { cell, rowspan: rowSpanToSet } );
 			}
 
-			const headingRows = table.getAttribute( 'headingRows' ) || 0;
+			const isCellOverlappingRemovedRows = row < first && lastRowOfCell >= first;
 
-			if ( headingRows && first < headingRows ) {
-				const newRows = getNewHeadingRowsValue( first, last, headingRows );
+			if ( isCellOverlappingRemovedRows ) {
+				const rowspanAdjustment = lastRowOfCell >= last ? rowsToRemove : first - row;
```
[the remark this answers is not in the document]
Comment (jodator) [reply src/tableutils.js]: You were right. The code wasn't clear enough - even for me.

I've refactored it once again and moved it outside main method execution. Hopefully this is can be parsed a bit better.

As for other cases we will probably see :crossed\_fingers:.

[the remark this answers is not in the document]
Comment (jodator) [reply src/tableutils.js]: Change is here: c4bfddb.

File: tests/tableutils.js [hunk 6]
```diff
@@ -1084,6 +1130,48 @@ describe( 'TableUtils', () => {
...
+
+				tableUtils.removeRows( root.getChild( 0 ), { at: 2, rows: 2 } );
+
+				// +----+----+----+----+----+
+				// | 00 | 01 | 02 | 03 | 04 |
+				// +----+    +    +    +    +
+				// | 10 |    |    |    |    |
+				// +----+----+----+----+    +
+				// | 40 | 41 | 42 | 43 |    |
+				// +----+----+----+----+----+
```
Comment (niegowski) [inline tests/tableutils.js:1165]: Missing 5th row in picture.

File: src/tableutils.js [hunk 7]
```diff
@@ -789,6 +765,69 @@ function updateHeadingRows( table, first, last, model, batch ) {
...
+//      +    +    +----+    +    +
+//      |    |    | 32 |    |    | <-- removed row
+//      +----+    +    +----+    +
+//      | 40 |    |    | 43 |    |
+//      +----+----+----+----+----+
+//
+// In a table above:
+// - cells to trim: '02', '03' & '04'.
+// - cells to move: '21' & '32'.
+function getCellsToMoveAndTrimOnRemoveRow( table, last, first, rowsToRemove ) {
```
Comment (niegowski) [inline src/tableutils.js:789]: Order of arguments is confusing, why not like this:
```
function getCellsToMoveAndTrimOnRemoveRow( table, first, last )
```
Also `rowsToRemove` seems redundant.

Comment (jodator) [reply src/tableutils.js]: Yeah - I don't wanted to recalculate this but it is in one place only and with a comment so I removed that.

The order of arguments was also fixed.

**Comment by jodator** [comment]
@oleq So the main review is done by @niegowski.

File: src/tableutils.js [hunk 8]
```diff
@@ -283,27 +283,35 @@ export default class TableUtils extends Plugin {
...
-
-			if ( headingRows && first < headingRows ) {
-				const newRows = getNewHeadingRowsValue( first, last, headingRows );
-
-				// Must be done after the changes in table structure (removing rows).
-				// Otherwise the downcast converter for headingRows attribute will fail. ckeditor/ckeditor5#6391.
-				model.enqueueChange( writer.batch, writer => {
-					updateNumericAttribute( 'headingRows', newRows, table, writer, 0 );
-				} );
+			// 2c. Update cells from rows above that overlaps removed section. Similar to step 2 but does not involve moving cells.
```
Comment (oleq) [inline src/tableutils.js:308]:
```suggestion
			// 2c. Update cells from rows above that overlap removed section. Similar to step 2 but does not involve moving cells.
```

File: src/tableutils.js [hunk 9]
```diff
@@ -730,60 +738,123 @@ function breakSpanEvenly( span, numberOfCells ) {
...
+		// Must be done after the changes in table structure (removing rows).
+		// Otherwise the downcast converter for headingRows attribute will fail. ckeditor/ckeditor5#6391.
+		model.enqueueChange( batch, writer => {
+			updateNumericAttribute( 'headingRows', newRows, table, writer, 0 );
+		} );
 	}
+}
+
+// Finds cells that will be:
+// - trimmed - Cells that are "above" removed rows sections and overlaps removed section - their rowspan must be trimmed.
```
Comment (oleq) [inline src/tableutils.js:769]:
```suggestion
// - trimmed - Cells that are "above" removed rows sections and overlap the removed section - their rowspan must be trimmed.
```

File: src/tableutils.js [hunk 10]
```diff
@@ -730,60 +738,123 @@ function breakSpanEvenly( span, numberOfCells ) {
...
+		// Otherwise the downcast converter for headingRows attribute will fail. ckeditor/ckeditor5#6391.
+		model.enqueueChange( batch, writer => {
+			updateNumericAttribute( 'headingRows', newRows, table, writer, 0 );
+		} );
 	}
+}
+
+// Finds cells that will be:
+// - trimmed - Cells that are "above" removed rows sections and overlap the removed section - their rowspan must be trimmed.
+// - moved - Cells from removed rows section might stick out of. Such cells are moved to a next row after a removed section.
```
Comment (oleq) [inline src/tableutils.js:770]:
```suggestion
// - moved - Cells from removed rows section might stick out of. These cells are moved to the next row after a removed section.
```

**Comment by oleq** [comment]
Can you reproduce this @jodator?

**Comment by jodator** [comment]
> The only concern as commented: [#303 (comment)](https://github.com/ckeditor/ckeditor5-table/pull/303#issuecomment-614555972)

@oleq It looks like it is downcast error - I'll handle this in ckeditor/ckeditor5#6437.
